City-of-Helsinki/drupal-helfi-rekry
# UHF-5293:

# [Helbit migration](https://helsinkisolutionoffice.atlassian.net/browse/UHF-5292)

Add migrations for Helbit data

## What was done?

Added functionality to migrate data from Helbit API to Drupal. This includes:
- Installed `migrate_plus` module to help with migrations
- Migration YAMLS
- New taxonomies (employment, employment type, task area)
- Some custom code to help migrating the data

Custom stuff:
Added a new Deriver class for Helbit migrations. This makes it possible to run migrations with some extra settings from the command line. For example, the id for job listing migrations is `helfi_rekry_jobs`. With the deriver this can be run as:
- `helfi_rekry_jobs:all`
- `helfi_rekry_jobs:changed`
- `helfi_rekry_jobs:all_sv`
- `helfi_rekry_jobs:changed_sv`
- `helfi_rekry_jobs:all_en`
- `helfi_rekry_jobs:changed_en`

... thus adding the option of importing only job listings changed in the past 24hrs instead of importing all of them.

Added an event listener class `Drupal\helfi_rekry_content\EventSubscriber\ImageImportSubscriber`. This creates a new directory for images of job listings, keeping them in a separate location. It also creates a media entity from the imported images to be used in job listing nodes.

Added a new data parser plugin `Drupal\helfi_rekry_content\Plugin\migrate_plus\data_parser\HelbitJson`. It seems that if no jobs that match parameters are found, Helbit api doesn't return an empty array or anything, causing data parser plugin to throw errors. The custom parser checks if the response is valid by the status it gives and throws errors only if it is not OK.

## How to test?

- Add `HELBIT_CLIENT_ID` variable to your .env. Value for this can be found in the jira ticket: (not writing it here since repos are public). Do `docker-compose up -d --build app` and verify that the variable is loaded correctly with `make shell` and `printenv`
- Checkout branch. Run `make fresh`.
- `make shell` into the container. Run `drush mim helfi_rekry_images:all` (this might take a while). Once the migration is under way you should start seeing content in [files](https://helfi-rekry.docker.so/fi/admin/content/files) and [media](https://helfi-rekry.docker.so/fi/admin/content/media). The migration should finish with mostly successful creations and a much lower amount of ignored ones. (Latest run I got 560 total with 466 successful and 94 ignored - ignored ones are job listings without image uri value).
- Run `drush mim helfi_rekry_organizations:all; drush mim helfi_rekry_task_areas:all`. Then check that you're getting taxonomy data for [organizations](https://helfi-rekry.docker.so/fi/admin/structure/taxonomy/manage/organization/overview) and [task areas](https://helfi-rekry.docker.so/fi/admin/structure/taxonomy/manage/task_area/overview).
- Run `drush mim helfi_rekry_organizations:all_sv; drush mim helfi_rekry_organizations:all_en; drush mim helfi_rekry_task_areas:all_sv; drush mim helfi_rekry_task_areas:all_en`. The previously imported terms should now have translations in all 3 enabled languages.
- Run `drush mim helfi_rekry_jobs:all` (this might take a while). As the migration is under way, you should start seeings [job listing nodes](https://helfi-rekry.docker.so/fi/admin/content?title=&type=job_listing&status=All&langcode=All). Check a few nodes and see that the fields are gettings populated. None of the nodes should have `field_video` value at this point. You should be able to find values for all other fields within a few nodes
- Try running some of these migrations: 
 - `drush mim helfi_rekry_jobs:changed`
 - `drush mim helfi_rekry_jobs:changed_sv`
 - `drush mim helfi_rekry_jobs:changed_en`
 some of them might migrate nothing, but none of them should error.
